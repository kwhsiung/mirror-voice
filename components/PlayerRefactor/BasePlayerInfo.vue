<template>
  <div class="info">
    <nuxt-link
      class="info__title info__title--truncate"
      :to="to"
      v-text="text"
    />
    <div class="info__time info__title--fixed-width">
      <span
        class="time-current"
        v-text="timeCurrentFormatted"
      />
      <span
        class="separator"
        v-text="separator"
      />
      <span
        class="time-total"
        v-text="timeTotalFormatted"
      />
    </div>
  </div>
</template>

<script>
export default {
  props: {
    to: {
      type: String,
      default: '/'
    },
    text: {
      type: String,
      default: ''
    },
    timeTotal: {
      type: Number,
      default: 0,
      validator(value) {
        return value >= 0
      }
    },
    timeCurrent: {
      type: Number,
      default: 0,
      validator(value) {
        return value >= 0
      }
    }
  },
  data() {
    return {
      separator: '/'
    }
  },
  computed: {
    timeTotalFormatted() {
      return this.$secondsToHms(this.timeTotal)
    },
    timeCurrentFormatted() {
      return this.$secondsToHms(this.timeCurrent)
    }
  }
}
</script>

<style lang="stylus" scoped>
.info
  display flex
  justify-content space-between
  font-size 12px
  &__title
    line-height 1
    color #e5e5e5
    display block
    width 0px
    flex 1 1 auto
    &--truncate
      white-space nowrap
      overflow hidden
      text-overflow ellipsis
  &__time
    line-height 1
    color #e5e5e5
    &--fixed-width
      min-width max-content
      margin 0 0 0 5px

.separator, .time-total
  color #7d7d7d

@media (max-width 768px)
  .info
    font-size 11px

  .separator, .time-total
    display none
</style>
