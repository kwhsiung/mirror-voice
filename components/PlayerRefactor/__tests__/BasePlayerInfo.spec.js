import _ from 'lodash'
import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import BasePlayerInfo from '../BasePlayerInfo.vue'
import secondsToHms from '~/plugins/util/secondsToHms'

// customizer to overwrite properties if the source object properties are an empty object or an array, like overwrite store object/array by empty object/array.
const customizer = (objValue, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue
  }
}
const createWrapper = overrides => {
  const defaultMountingOptions = {
    stubs: {
      NuxtLink: RouterLinkStub
    },
    mocks: {
      $secondsToHms: secondsToHms
    }
  }
  return shallowMount(
    BasePlayerInfo,
    _.mergeWith(defaultMountingOptions, overrides, customizer)
  )
}

describe('BasePlayerInfo title', () => {
  test('render NuxtLink with "to" received from "to" props', () => {
    const to = '/test/path'
    const wrapper = createWrapper({
      propsData: {
        to
      }
    })
    const link = wrapper.find(RouterLinkStub)

    expect(link.props().to).toBe(to)
  })
  test('render text received from "text" props', () => {
    const text = 'test title'
    const wrapper = createWrapper({
      propsData: {
        text
      }
    })
    const link = wrapper.find(RouterLinkStub)

    expect(link.text()).toBe(text)
  })
})

describe('BasePlayerInfo time', () => {
  test('render formatted "timeTotal" received from props', () => {
    const timeInSecond = 1200
    const wrapper = createWrapper({
      propsData: {
        timeTotal: timeInSecond
      }
    })
    const timeTotal = wrapper.find('.time-total')

    expect(timeTotal.text()).toBe('00:20:00')
  })
  test('render formatted "timeCurrent" received from props', () => {
    const timeInSecond = 1200
    const wrapper = createWrapper({
      propsData: {
        timeCurrent: timeInSecond
      }
    })
    const timeCurrent = wrapper.find('.time-current')

    expect(timeCurrent.text()).toBe('00:20:00')
  })
})
