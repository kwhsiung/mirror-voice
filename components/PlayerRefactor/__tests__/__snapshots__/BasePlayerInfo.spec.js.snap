// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders component correctly if we provide efficient props 1`] = `
<div
  class="info"
>
  <a
    class="info__title info__title--truncate"
  >
    test text
  </a>
   
  <div
    class="info__time info__title--fixed-width"
  >
    <span
      class="time-current"
    >
      00:02:03
    </span>
     
    <span
      class="separator"
    >
      /
    </span>
     
    <span
      class="time-total"
    >
      03:25:45
    </span>
  </div>
</div>
`;
